File: rust/examples/linux/power/ade7953/complete/src/main.rs
use linux_embedded_hal::I2cdev;
use periph::chips::power::Ade7953Minimal;
use embedded_hal::delay::DelayNs;
use std::time::Duration;

struct Delay;

impl DelayNs for Delay {
    fn delay_ms(&mut self, ms: u32) {
        std::thread::sleep(Duration::from_millis(ms as u64));
    }
}

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);
    let addr: u8 = std::env::var("I2C_ADDR")
        .ok()
        .and_then(|v| u8::from_str_radix(v.trim_start_matches("0x"), 16).ok())
        .unwrap_or(0x38);

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let mut delay = Delay;
    let mut ade = Ade7953Minimal::new(dev, addr, 251.0, 30.0, &mut delay).expect("init ADE7953");  // Create ADE7953 driver, (i2c, addr, voltage_gain=V/V, current_gain=A/V, delay)

    println!("version: 0x{:02X}", ade.version().expect("version"));     // Read silicon version, () → u8
    println!("V={:.2}",   ade.voltage().expect("voltage"));              // Read bus voltage, () → f32 V
                                                                            // converts raw VRMS to volts using voltage_gain
    println!("I_a={:.3}", ade.current().expect("current"));              // Read load current, () → f32 A
                                                                            // converts raw IRMSA to amperes using current_gain
    println!("P_a={:.2}", ade.active_power().expect("active_power"));     // Read active power, () → f32 W
                                                                            // converts raw AWATT (instantaneous, 6.99 kHz) to watts
    println!("E_a={:.4}", ade.active_energy().expect("active_energy"));   // Read active energy, () → f32 Wh
                                                                            // converts raw AENERGYA accumulated LSBs to watt-hours
    println!("PF={:.3}",  ade.power_factor().expect("power_factor"));     // Read power factor, () → f32 ratio
                                                                            // converts raw PFA (1 LSB = 2^-15) to a −1.0..+1.0 ratio
    println!("f={:.2}",    ade.line_frequency().expect("line_frequency")); // Read line frequency, () → f32 Hz
    println!("Q={:.2}",   ade.reactive_power().expect("reactive_power")); // Read reactive power, () → f32 VAR
    println!("S={:.2}",   ade.apparent_power().expect("apparent_power")); // Read apparent power, () → f32 VA

    ade.configure_overvoltage(260.0).expect("overvoltage");              // Configure overvoltage, (threshold) → ()
                                                                            // threshold in volts (same scale as voltage())
    ade.configure_overcurrent(40.0).expect("overcurrent");               // Configure overcurrent, (threshold) → ()
                                                                            // threshold in amperes; applies to BOTH current channels

    ade.reset(&mut delay).expect("reset");                                // Software reset, () → ()
                                                                            // waits 110 ms then re-runs the mandatory power-up sequence
}

File: cpp/examples/linux/power/ADE7953/complete/main.cpp
#include <cstdio>
#include "I2CConnectionLinux.h"
#include "ADE7953.h"

#ifndef TEST_I2C_BUS
#define TEST_I2C_BUS 1
#endif
#ifndef TEST_ADDR
#define TEST_ADDR 0x38
#endif

static const float VOLTAGE_GAIN   = 251.0f;
static const float CURRENT_GAIN_A = 30.0f;
static const float CURRENT_GAIN_B = 30.0f;

int main() {
    I2CConnectionLinux conn(TEST_I2C_BUS, TEST_ADDR);
    ADE7953Full ade(conn, VOLTAGE_GAIN, CURRENT_GAIN_A);             // Create ADE7953 driver, (connection, voltage_gain=V/V, current_gain=A/V, bus_type='i2c')

    std::printf("version: 0x%02X\n", ade.version());                  // Read silicon version, () → 0x..
                                                                       // returns the silicon revision

    std::printf("V=%.2f\n",   ade.voltage());                        // Read bus voltage, () → V
                                                                       // converts raw VRMS to volts using voltage_gain
    std::printf("I_a=%.3f\n", ade.current());                        // Read load current, () → A
                                                                       // converts raw IRMSA to amperes using current_gain
    std::printf("P_a=%.2f\n", ade.activePower());                    // Read active power, () → W
                                                                       // converts raw AWATT (instantaneous, 6.99 kHz) to watts
    std::printf("E_a=%.4f\n", ade.activeEnergy());                   // Read active energy, () → Wh
                                                                       // converts raw AENERGYA accumulated LSBs to watt-hours
    std::printf("PF=%.3f\n",  ade.powerFactor());                    // Read power factor, () → ratio
                                                                       // converts raw PFA (1 LSB = 2^-15) to a −1.0..+1.0 ratio
    std::printf("f=%.2f\n",    ade.lineFrequency());                  // Read line frequency, () → Hz

    ade.configureChannelB(CURRENT_GAIN_B);                            // Set Channel B calibration, (current_gain_b) → none
    std::printf("I_b=%.3f\n", ade.currentB());                        // Read Current Channel B, () → A
    std::printf("P_b=%.2f\n", ade.activePowerB());                    // Read Channel B active power, () → W

    ade.setActiveEnergyMode('a', 0);                                  // Set active-energy mode A, (channel, mode) → none
                                                                       // mode = 0 (normal) | 1 (positive-only) | 2 (absolute)
    ade.setReactiveEnergyMode('a', 0);                                // Set reactive-energy mode A, (channel, mode) → none
                                                                       // mode = 0 (normal) | 1 (antitamper) | 2 (absolute)

    ade.setPga('a', 1);                                               // Write PGA Channel A, (channel, gain) → none
                                                                       // gain 1, 2, 4, 8, 16 (+22 valid only for Channel A)
    ade.setPga('v', 1);                                               // Write PGA voltage channel, (channel, gain) → none
    ade.setPhaseCalibration('a', 0.0f);                               // Write phase calibration A, (channel, delay_s) → none
                                                                       // negative delay_s advances; range ±383 × 1.117 µs
    ade.setGainCalibration(0x282, 0x400000);     // Write active-power gain A, (reg, value) → none
                                                                       // 0x400000 = unity; valid range 0x200000..0x600000
    ade.setOffsetCalibration(0x289, 0);        // Write active-power offset A, (reg, value) → none
                                                                       // signed 24-bit offset

    std::printf("checksum: 0x%08X\n", (unsigned)ade.checksum());     // Read CRC/checksum, () → u32
    ade.enableChecksum(true);                                         // Enable CRC/checksum, (enabled) → none

    ade.configureOvervoltage(260.0f);                                 // Configure overvoltage, (threshold) → none
                                                                       // threshold in volts (same scale as voltage())
    ade.configureOvercurrent(40.0f);                                  // Configure overcurrent, (threshold) → none
                                                                       // threshold in amperes; applies to BOTH current channels

    ade.reset();                                                      // Software reset, () → none
                                                                       // waits 110 ms then re-runs the mandatory power-up sequence
    return 0;
}